### PR TITLE
Cassandra Binary Service and tests

### PR DESCRIPTION
**Pull request by gregjan**
Cassandra Binary Service and tests.

### CODE REVIEW

**Comment by gregjan** [comment]
Travis seems to be stuck on an earlier version of the Trellis snapshots? I made changes today (in this PR) that got all of the interfaces up to date with respect to Trellis 0.7

**Comment by acoburn** [comment]
@gregjan you might want to update the pom files to point to Trellis version 0.7.0. Right now, it's using the 0.7.0-SNAPSHOT, which may be a little out of date.

File: src/main/java/edu/si/trellis/cassandra/CassandraResourceService.java [hunk 1]
```diff
@@ -172,15 +167,11 @@ public String generateIdentifier() {
         return execute(immutableDataInsert);
     }
 
-    @Override
-    public Future<Boolean> create(IRI id, Session session, IRI ixnModel, Dataset dataset, IRI container,
-                    Binary binary) {
+    public Future<Boolean> create(IRI id, Session session, IRI ixnModel, Dataset dataset) {
```
Comment (ajs6f) [inline src/main/java/edu/si/trellis/cassandra/CassandraResourceService.java]: I'm confused-- are you not writing against the latest Trellis API?

File: src/main/java/edu/si/trellis/cassandra/CassandraBinaryService.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+		readRangeStatement = session.prepare(READ_RANGE_QUERY);
+		deleteStatement = session.prepare(DELETE_QUERY);
+	}
+
+	@Override
+	public Optional<InputStream> getContent(IRI identifier, List<Range<Integer>> ranges) {
+		requireNonNull(ranges, "Byte ranges may not be null");
+		try {
+			if (ranges.isEmpty()) {
+				return Optional.ofNullable(readAll(identifier));
```
Comment (ajs6f) [inline src/main/java/edu/si/trellis/cassandra/CassandraBinaryService.java]: Does `readAll` ever return `null`? If the binary doesn't exist, I think it just returns an empty `InputStream`...

**Comment by ajs6f** [comment]
@acoburn Right now, @gregjan has to spin threads to impl `getContent`, and he would have to do something like that to return the current `Optional<InputStream>`. I don't like stacking up types too high, but if that were `Optional<Future<InputStream>>` he could return a `Future` and let thread management occur on the other side, which is more like what we want, I think. (@gregjan You could use an asynchronous C* query for that-- I've done it elsewhere.)) In the simple case, of course, the HTTP layer can just resolve that `Future` right away to deliver bits on the request thread. I think this would let @gregjan get rid of the `Piped` streams as well.

**Comment by acoburn** [comment]
Should the return type be `Optional<Future<InputStream>>` or `Future<Optional<InputStream>>`?

**Comment by ajs6f** [comment]
[`Optional<Future<InputStream>>` or `Future<Optional<InputStream>>`?](https://www.youtube.com/watch?v=TJ9f2rnjB84)

Well, I guess one question is what happens if there is a bitstream but the retrieval blows up before getting one bit? There's nowhere elegant to put an exception but to rethrow it. So maybe even `CompletableFuture<Optional<InputStream>`, or am I getting too cute for my own good?

**Comment by ajs6f** [comment]
Orrrrr....
Just `CompletableFuture<InputStream>` where a missing bitstream completes the `CompletableFuture` with a specific `MissingBitstreamException` which is then a 404 at the HTTP level?

**Comment by acoburn** [comment]
I rather like the simplicity of `CompletableFuture<InputStream>`.

**Comment by ajs6f** [comment]
Would you want to revisit `ResourceService` and related types, which have several `Future<Boolean>` return types and `Optional<Resource> get(...)`? `::get` could very well be a `Future` of some kind.

**Comment by acoburn** [comment]
There is a huge value to consistency, so yes, that would definitely be something to revisit.

**Comment by ajs6f** [comment]
Okay, so this all raises a new question for me, which is how to order API changes and this PR? @acoburn, do you think we could make these API changes quickly? If so, I would hold this PR and either @gregjan or I can add a commit to update it before merging.If not, I'd rather merge this one, keep @gregjan moving, and then I can go back and update when we do change the API.

**Comment by acoburn** [comment]
I'd suggest merging this now, and deal with the API changes later. My main thinking is that there are a number of features that I'd like to go into Trellis 0.8, and I don't want @gregjan to be waiting a few weeks before I get all the ducks in a row to cut a release.

**Comment by ajs6f** [comment]
Okay, cool. @gregjan can you make the `pom.xml` update @acoburn describes above and let's get this merged!

**Comment by acoburn** [comment]
I made a ticket for the API changes here: https://github.com/trellis-ldp/trellis/issues/148

**Comment by ajs6f** [comment]
@gregjan Are you going to have a chance to update this PR today or tomorrow, or should we put it on hold until next week? I may be able to try to get it mergable myself, but that's iffy...

File: src/main/java/edu/si/trellis/cassandra/CassandraResourceService.java (hunk 1 above)
Comment (gregjan) [reply src/main/java/edu/si/trellis/cassandra/CassandraResourceService.java]: I guess that the SNAPSHOT modifier was holding be back at a previous commit. Will bring it up to date.

**Comment by ajs6f** [comment]
Did you mean to close this, @gregjan ?

**Comment by ajs6f** [comment]
Ah, you are at #3 , gotcha, see you over there.

**Comment by gregjan** [comment]
Yeah, just made the new one. Probably should have kept the old one, but
didn't see the option to add a commit to it..

**Comment by ajs6f** [comment]
That's cool. For the future, could you not have pushed more commits to the branch from which this PR came (`gregjan:master`)? That would normally auto-add them to the PR.

**Comment by gregjan** [comment]
aha, yeah that's right! I am so rusty.

**Pull request closed without merging**